File: code_combat/4_Backwoods_Forest/210-Defend_the_Garrison/defeatGar.js
// Defeat all of the attacking ogres.
// Use flags to move away from dangerous ogres.

while (true) {
    var flag = hero.findFlag();
    if (flag)
        hero.pickUpFlag(flag);
    var enemy = hero.findNearestEnemy();
    if (enemy) {
        if (hero.canCast("chain-lightning")) hero.cast("chain-lightning", enemy);
        else if (hero.isReady("electrocute")) hero.electrocute(enemy);
        else if (hero.isReady("cleave") && (hero.distanceTo(enemy) < 5)) hero.cleave(enemy);
        else if (hero.isReady("bash")) hero.bash(enemy);
        else hero.shield();
    }
}

File: projects/code_combat/4_Backwoods_Forest/210-Defend_the_Garrison/defeat_gar.py
# Unholy Tome 5 to "Poison Cloud", "Drain Life", "Raise Dead", "Fear", "Summon Undead"
# Emperor's ring to "Chain Lightning"
# Vine Staff to "Summon Burl"

while True:
    enemy = hero.findNearestEnemy()
    flag = hero.findFlag()

    if flag:
        hero.pickUpFlag(flag)

    if hero.canCast("summon-burl"):
        hero.cast("summon-burl")
    if hero.canCast("summon-undead"):
        hero.cast("summon-undead")

    if enemy:
        distance = hero.distanceTo(enemy)

        if distance <= 30:
            if hero.canCast("chain-lightning"):
                hero.cast("chain-lightning", enemy)
            elif distance > 15 and hero.canCast("poison-cloud"):
                hero.cast("poison-cloud", enemy)
            elif distance <= 25 and hero.canCast("fear"):
                hero.cast("fear", enemy)
            elif distance <= 15 and hero.canCast("drain-life"):
                hero.cast("drain-life", enemy)
            else:
                hero.attack(enemy)
        elif distance < 45:
            hero.attack(enemy)

    if hero.canCast("raise-dead"):
        hero.cast("raise-dead")
